configure.ac

```
@@ -202,6 +202,61 @@ AS_CASE([$with_gc],
     )
 AC_MSG_RESULT([$with_gc])
 
+dnl
+dnl User setting: native thread-local storage (off by default)
+dnl
+
+AC_ARG_WITH([native-tls],
+    [AS_HELP_STRING([--with-native-tls],
+        [use native thread-local storage implementation])],
+        [],
+        [with_native_tls=no])
+AC_MSG_CHECKING([whether to use native tls])
+AS_IF([test "x$with_native_tls" == "xyes"], [
+        AC_COMPILE_IFELSE(
+            [AC_LANG_SOURCE([[
+#if defined(__APPLE__) && defined(__x86_64__)
+#error FAIL
+#endif
+        ]])],
+            [
+                AC_DEFINE([USE_NATIVE_TLS], [1], [define as 1 if using native TLS])
+                AC_MSG_RESULT([yes])
+            ],
+            [
+                AC_DEFINE([USE_PTHREAD_TLS], [1], [define as 1 if using pthread_getspecific])
+                AC_MSG_RESULT([pthread])
+            ]
+        )
+        AC_RUN_IFELSE(
+            [AC_LANG_PROGRAM([[
+#include <pthread.h>
+#include <string.h>
+]],[[
+#if defined(__APPLE__) && defined(__x86_64__)
+    static char asm_code[] = {
+        // movq %gs:(,%rdi,8), %rax
+        // retq
+        0x65, 0x48, 0x8b, 0x04, 0xfd, 0x00, 0x00, 0x00, 0x00, 0xc3
+    };
+    if (memcmp((void *)pthread_getspecific, asm_code, sizeof(asm_code)) != 0) {
+        return 1;
+    }
+    return 0;
+#else
+#error FAIL
+#endif
+            ]])],
+            [AC_DEFINE([USE_MACOS_PTHREAD_TLS_ASM], [1],
+                [define as 1 if macOS assembly implementation supported])],
+            [],
+            [AC_MSG_NOTICE(["Don't support cross-compiling for HPC_GAP with native TLS"])])
```

Shouldn't this only be printed if cross compilation is being used? Also, the message is a bit backwards, how about this:
```suggestion
            [AC_MSG_NOTICE(["Cross compilation of HPC_GAP with native TLS is not supported"])])
```

But actually, if I read your code right, isn't this only executed if `with_native_tls` is *not* set to `yes`? I.e. we don't use "native" TLS?! Though then the macOS assembly check also seems misplaced?!? Maybe I skipped the wrong number of opening and closing brackets... Perhaps you could briefly summarize, in words, what the detection logic here is supposed to do, then I could make a concrete suggestion how to tweak the code to make it clearer.

---

This is the branch of `AC_RUN_IFELSE()` that is taken if cross-compiling is being done, in which case we can't do the test (it's the third branch after the "then" and "else" branches). That said, I've modified the output to simply add a message result rather than dissecting it further for now.

---

Adding a comment line like `dnl the following checks are only executed when cross compiling` would also help a lot in making this code more understandable.

---

Good point. I'll probably do another pass to break down the logic more cleanly, as there's a lot of nesting going on.